### PR TITLE
Add callback to indicate when joiner operation completes.

### LINKED ISSUE #820: No Notification for Commissioning Join Complete
OpenThread exposes the Joiner API `otJoinerStart` to start a join, but without polling logic, there is no way for the client to know when the joining logic has succeeded or failed. We should at least support a flag to indicate to clients registered with `otStateChangedCallback` when the joining process is completed.

### PR DESCRIPTION
**Pull request by jwhui**
Resolves #820.

Note: the Windows APIs are simply stubbed out by passing `NULL` for the callback.

### CODE REVIEW

File: src/core/meshcop/joiner.cpp [hunk 1]
```diff
@@ -340,6 +399,8 @@ void Joiner::SendJoinerEntrustResponse(const Coap::Header &aRequestHeader,
     memset(&responseInfo.mSockAddr, 0, sizeof(responseInfo.mSockAddr));
     SuccessOrExit(error = mCoapServer.SendMessage(*message, responseInfo));
 
+    mState = kStateJoined;
```
Comment (nibanks) [inline src/core/meshcop/joiner.cpp]: From what I can tell, it still waits for the `HandleTimer` before triggering the callback. Is that correct? If so, could you explain why?

File: examples/drivers/windows/otLwf/iocontrol.c [hunk 2]
```diff
@@ -5270,8 +5270,14 @@ otLwfIoCtl_otJoinerStart(
...
-        status = ThreadErrorToNtstatus(otJoinerStart(
-            pFilter->otCtx, (const char*)aConfig->PSKd, (const char*)aConfig->ProvisioningUrl));
+        status = ThreadErrorToNtstatus(
+            otJoinerStart(
+                pFilter->otCtx,
+                (const char*)aConfig->PSKd,
+                (const char*)aConfig->ProvisioningUrl,
+                NULL,
+                NULL)
+            );
```
Comment (nibanks) [inline examples/drivers/windows/otLwf/iocontrol.c]: Could you just put a TODO comment at all these places that need to be finished up?

File: src/core/meshcop/dtls.hpp [hunk 3]
```diff
@@ -93,15 +102,17 @@ class Dtls
...
+     * @param[in]  aReceiveHandler    A pointer to the receive handler.
+     * @param[in]  aSendHandler       A pointer to the send handler.
+     * @param[in]  aContext           A pointer to application-specific context.
      *
-     * @retval kThreadError_None  Successfully started the DTLS service.
+     * @retval kThreadError_None      Successfully started the DTLS service.
      *
      */
-    ThreadError Start(bool aClient, ReceiveHandler aReceiveHandler, SendHandler aSendHandler, void *aContext);
+    ThreadError Start(bool aClient, ConnectedHandler, ReceiveHandler aReceiveHandler, SendHandler aSendHandler,
```
Comment (rlubos) [inline src/core/meshcop/dtls.hpp]: Minor: aConnectedHandler is missing in the declaration.

Comment (jwhui) [inline src/core/meshcop/dtls.hpp]: Fixed, thanks!

**Pull request merged**